src/relay/pass/kind_check.cc

```diff
@@ -20,12 +20,71 @@ namespace tvm {
 namespace relay {
 
 using namespace tvm::runtime;
+using Kind = TypeParamNode::Kind;
 
 struct KindChecker : TypeVisitor<> {
   bool valid;
 
   KindChecker() : valid(true) {}
 
+  // checks if t is an incomplete node of kind k or a type param of kind k
+  bool MatchKind(const Type& t, Kind k) {
+    if (const IncompleteTypeNode *tv = t.as<IncompleteTypeNode>()) {
+      return tv->kind == k;
+    }
+
+    if (const TypeParamNode *tp = t.as<TypeParamNode>()) {
+      return tp->kind == k;
+    }
+
+    return false;
+  }
+
+  bool IsTypeKind(const Type& t) {
+    if (MatchKind(t, Kind::kType)) {
+      return true;
+    }
+
+    return t.as<TensorTypeNode>() || t.as<BaseTensorTypeNode>() || t.as<TupleTypeNode>() || t.as<FuncTypeNode>();
```

Comment:
Not entirely sure why this happened, but my tests failed if I tried to cast a tensor type to a BaseTensorTypeNode, but not if I cast them to a TensorTypeNode. Is this because a TensorType instance doesn't know that its representation (the one you `get()`) inherits from BaseTensorTypeNode? In any case, that doesn't seem like the desired behavior but I don't know what would be preferred.